glide-core/redis-rs/redis/src/commands/cluster_scan.rs

```diff
@@ -323,51 +367,73 @@ impl ScanState {
             return Ok(ScanState::create_finished_state());
         } else {
-            connection.get_address_by_slot(next_slot).await
+            connection
+                .get_address_by_slot(next_slot, &mut scanned_slots_map, allow_non_covered_slots)
+                .await?
         };
-        match new_address {
-            Ok(new_address) => {
-                let new_epoch = connection
-                    .get_address_epoch(&new_address)
-                    .await
-                    .unwrap_or(0);
-                let new_cursor = 0;
-                Ok(ScanState::new(
-                    new_cursor,
-                    scanned_slots_map,
-                    new_address,
-                    new_epoch,
-                    ScanStateStage::InProgress,
-                ))
-            }
-            Err(err) => Err(err),
+
+        if *new_address == NO_ADDRESSES_FOUND {
+            return Ok(ScanState::create_finished_state());
         }
+        let new_epoch = connection
+            .get_address_epoch(&new_address)
+            .await
+            .unwrap_or(0);
+        let new_cursor = 0;
+        Ok(ScanState::new(
+            new_cursor,
+            scanned_slots_map,
+            new_address,
+            new_epoch,
+            ScanStateStage::InProgress,
+        ))
     }
 }
 
 // Implement the [`ClusterInScan`] trait for [`InnerCore`] of async cluster connection.
-#[async_trait]
 impl<C> ClusterInScan for Core<C>
 where
     C: ConnectionLike + Connect + Clone + Send + Sync + 'static,
 {
-    async fn get_address_by_slot(&self, slot: u16) -> RedisResult<Arc<String>> {
-        let address = self
-            .get_address_from_slot(slot, SlotAddr::ReplicaRequired)
-            .await;
-        match address {
-            Some(addr) => Ok(addr),
-            None => {
-                if self.are_all_slots_covered().await {
-                    Err(RedisError::from((
+    async fn get_address_by_slot(
+        &self,
+        mut slot: u16,
+        scanned_slots_map: &mut SlotsBitsArray,
+        allow_non_covered_slots: bool,
+    ) -> RedisResult<Arc<String>> {
+        loop {
+            let address = self
+                .get_address_from_slot(slot, SlotAddr::ReplicaRequired)
+                .await;
+
+            match address {
+                Some(addr) => return Ok(addr),
+                None => {
+                    if allow_non_covered_slots {
+                        // Mark the current slot as scanned before moving to next
+                        let slot_index = slot as usize / BITS_PER_U64;
+                        let slot_bit = slot as usize % BITS_PER_U64;
+                        scanned_slots_map[slot_index] |= 1 << slot_bit;
+
+                        slot += 1;
+                        if slot == NUM_OF_SLOTS as u16 {
+                            return Ok(NO_ADDRESSES_FOUND.to_string().into());
+                        }
+                        continue;
```

Comment:
This code could potentially lead to duplicate slot scanning. Instead of simply incrementing to the next slot, we should find the next unscanned slot by checking the `scanned_slots_map`.

Reply:
true, it should be get next slot.